File: src/main/java/shopping/member/service/MemberCommandService.java
package shopping.member.service;

import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import shopping.auth.AuthService;
import shopping.member.domain.Member;
import shopping.member.repository.MemberRepository;
import shopping.member.service.dto.MemberLoginInput;
import shopping.member.service.dto.MemberRegisterInput;
import shopping.member.service.dto.TokenOutput;

@Service
@Transactional
@RequiredArgsConstructor
public class MemberCommandService {
    private final MemberRepository memberRepository;
    private final MemberQueryService memberQueryService;
    private final AuthService authService;

    public TokenOutput register(MemberRegisterInput input) {
        verifyEmail(input.email());
        Member saved = memberRepository.save(Member.builder()
                                                   .email(input.email())
                                                   .password(authService.encodePassword(input.password()))
                                                   .build());
        String token = authService.createToken(saved.getId());
        return new TokenOutput(token);
    }

    public TokenOutput login(MemberLoginInput input) {
        Member member = memberQueryService.getMemberByEmail(input.email());
        authService.verifyPassword(input.password(), member.getPassword());
        String token = authService.createToken(member.getId());
        return new TokenOutput(token);
    }

    private void verifyEmail(String email) {
        if (memberQueryService.existsByEmail(email)) {
            throw new IllegalArgumentException("이미 사용 중인 이메일입니다.");
        }
    }
}

File: src/main/kotlin/shopping/member/application/MemberCommandService.kt
package shopping.member.application

import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import shopping.global.exception.ApplicationException
import shopping.global.exception.ErrorCode
import shopping.member.application.command.MemberCreateCommand
import shopping.member.domain.Member

@Service
@Transactional
class MemberCommandService(
    private val memberCommandRepository: MemberCommandRepository,
    private val memberQueryRepository: MemberQueryRepository,
    private val passwordEncoder: PasswordEncoder
) {

    fun createMember(memberCreateCommand: MemberCreateCommand): Long {
        val member = memberCreateCommand.toEntity(passwordEncoder)

        if (isAlreadyExistEmail(member)) {
            throw ApplicationException(ErrorCode.DUPLICATED_REGISTER_EMAIL)
        }

        return memberCommandRepository.save(member).id
    }

    private fun isAlreadyExistEmail(member: Member): Boolean {
        return memberQueryRepository.existsByEmailAndNotDeleted(member.email)
    }

}
